### PR TITLE
[rework] hiding camera

### LINKED ISSUE #312: [Rework] Hiding and closet interaction
- Liat milanote untuk implement terbaru [https://app.milanote.com/1NzRxK116UBW1s/hiding?p=bOJzjVxeHMY](https://app.milanote.com/1NzRxK116UBW1s/hiding?p=bOJzjVxeHMY)

### PR DESCRIPTION
**Pull request by naufalhilmi12345**
### Main changes

- Rework camera when hiding

### Tasks List

- [ close #312 ]

### Attachments
can test this at naufal/hiding_test scene

https://user-images.githubusercontent.com/44135006/187450690-44d8d200-db18-484c-be8a-aa4665aed8b0.mp4

Docs link: https://docs.google.com/document/d/166j1tS_sLd9ZBvMeYCcn66V-PgR6o6AluzB1jvh0rRU/edit?usp=sharing

### CODE REVIEW

**Review by rafiftaris — CHANGES_REQUESTED** [review]
- enclose code within a region
- put docs above class
- public field use PascalCase
- move CinemachinePOVExtension to Scripts/Camera


In order to encapsulate the details of Closets from PlayerHidingState (PlayerHidingState only acts as a mediator to CinemachinePOVExtension), pass the whole ClosetCameraSettings directly to CinemachinePOVExtension and get the values in that class.

**Review by rafiftaris — CHANGES_REQUESTED** [review]
- remove on camera live event on the Camera Hiding gameobject. make sure to remove it on the prefab too
image

File: Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+using System;
+using UnityEngine;
+
+/*
+ * Class to keep value for camera
+ */
+[Serializable]
+public class ClosetCameraSetting : MonoBehaviour
```
Comment (InShineStudios) [inline Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs:8]: Rename class: Hiding Camera Configs

File: Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+{
+    #region Variables
+    [Header("Camera Configs")]
+    [Tooltip("Limit for X rotation, 2D Vector where X is min degree and Y is max degree")]
+    public Vector2 ClampAngleX;
```
Comment (InShineStudios) [inline Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs:13]: Rename: Clamp Down Up

File: Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    [Tooltip("Limit for Y rotation, 2D Vector where X is min degree and Y is max degree")]
+    public Vector2 ClampAngleY;
```
Comment (InShineStudios) [inline Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs:15]: Clamp Left Right

File: Assets/Astralization/Scripts/Camera/CinemachinePOVExtension.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+    #region SetGet
+    public void SetMouseDeltaInput(Vector2 deltaInput)
+    {
+        _deltaInput = deltaInput;
+    }
+
+    public void SetClosetCameraSetting(ClosetCameraSetting closetCameraSetting)
+    {
+        _closetCameraSetting = closetCameraSetting;
```
Comment (InShineStudios) [inline Assets/Astralization/Scripts/Camera/CinemachinePOVExtension.cs:24]: Add Vector3 variable currentCameraDirection, save the original direction from Configs, use this variable to be adjusted on PostPipelineStageCallback

File: Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs [hunk 5]
```diff
@@ -0,0 +1,31 @@
+using System;
+using UnityEngine;
+
+/*
+ * Class to keep value for camera
```
Comment (InShineStudios) [inline Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs:5]: for hiding camera configs

File: Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    public Vector2 ClampAngleX;
+    [Tooltip("Limit for Y rotation, 2D Vector where X is min degree and Y is max degree")]
+    public Vector2 ClampAngleY;
+    [Tooltip("3D Vector indicates the camera direction")]
+    public Vector3 StartingDirection;
+    [Tooltip("3D Vector indicates the camera position when start hiding")]
+    public Vector3 StartingPosition;
+    #endregion
+
+    #region Handler
```
Comment (InShineStudios) [inline Assets/Astralization/Scripts/Interactables/Closet/ClosetCameraSetting.cs:22]: Constructor

**Review by InShineStudios — CHANGES_REQUESTED** [review]

**Comment by rafiftaris** [comment]
Reminder: Push test sama bikin diagram

File: Assets/Astralization/Tests/InteractableTest/InteractableTest.cs [hunk 7]
```diff
@@ -77,19 +77,29 @@ public IEnumerator PlayerInteractableDetector_InteractWoodenDoor()
...
-        Vector3 calOffset = GameObject.Find("Closets/Model/RotateRight").GetComponentInChildren<Renderer>().bounds.center;
+        Vector3 calOffset = closets.GetComponentInChildren<Renderer>().bounds.center;
+        calOffset.y = 0;
         //Assert.Less(Utils.GeometryCalcu.GetDistance3D(calOffset,player.transform.position),1f); //TODO #312
         //Assert.True(anim.GetBool("isHiding"));
         Assert.AreEqual(_playerInput.currentActionMap.name, "Hiding");
 
+        // Closets camera
+        HidingCameraConfigs hidingCameraConfigs = closets.GetComponentInChildren<HidingCameraConfigs>();
+        Vector3 startingPosition = hidingCameraConfigs.StartingPosition;
```
Comment (rafiftaris) [inline Assets/Astralization/Tests/InteractableTest/InteractableTest.cs:97]: rename to hidingCameraStartingPosition

**Review by rafiftaris — CHANGES_REQUESTED** [review]
the test is good but please rename this variable for clarity

**Review by rafiftaris — APPROVED** [review]
looks great to me~

**Review by InShineStudios — APPROVED** [review]
Nicely done

**Review by RaiRiz — APPROVED** [review]

**Pull request merged**